functorch/csrc/BatchRulesModules.cpp

```diff
@@ -200,6 +200,66 @@ std::tuple<Tensor,Tensor> cudnn_convolution_backward_plumbing(const Tensor & sel
   return slow_fallback<Tensor,Tensor>(op, { self, grad_output, weight, padding, stride, dilation, groups, benchmark, deterministic, allow_tf32, output_mask });
 }
 
+std::tuple<Tensor,optional<int64_t>>
+grid_sample_batch_rule(const Tensor& input, optional<int64_t> input_bdim, const Tensor& grid, optional<int64_t> grid_bdim, int64_t interpolation_mode, int64_t padding_mode, bool align_corners) {
+  std::vector<int64_t> input_spec {0, 1, 2, 3};
+  std::vector<int64_t> grid_spec = input_spec;
+  std::vector<int64_t> out_spec = input_spec;
+
+  std::tuple<Tensor, optional<int64_t>> result;
+  if (input_bdim && !grid_bdim) {
+    auto new_input = reshape_dim_into(*input_bdim, input_spec[1], input);
+    auto out = at::grid_sampler(new_input, grid, interpolation_mode, padding_mode, align_corners);
+    out = reshape_dim_outof(out_spec[1], input.sizes()[*input_bdim], out);
+    result = std::make_tuple(out, out_spec[1]);
+  } else if (!input_bdim && grid_bdim) {
+    // grid of N(BH)W2 -> NC(BH)W or grid of N(BD)HBW3 -> NC(BD)HW
+    auto new_grid = reshape_dim_into(*grid_bdim, grid_spec[1], grid);
+    auto out = at::grid_sampler(input, new_grid, interpolation_mode, padding_mode, align_corners);
+    out = reshape_dim_outof(out_spec[2], grid.sizes()[*grid_bdim], out);
+    result = std::make_tuple(out, out_spec[2]);
+  } else if (input_bdim && grid_bdim) {
+    auto new_input = reshape_dim_into(*input_bdim, input_spec[0], input);
+    auto new_grid = reshape_dim_into(*grid_bdim, grid_spec[0], grid);
+    auto out = at::grid_sampler(new_input, new_grid, interpolation_mode, padding_mode, align_corners);
+    out = reshape_dim_outof(out_spec[0], input.sizes()[*grid_bdim], out);
+    result = std::make_tuple(out, out_spec[0]);
+  } else {
+    result = std::make_tuple(at::grid_sampler(input, grid, interpolation_mode, padding_mode, align_corners), nullopt);
+  }
+  return result;
+}
+
+std::tuple<Tensor,optional<int64_t>>
+cudnn_grid_sample_batch_rule(const Tensor& input, optional<int64_t> input_bdim, const Tensor& grid, optional<int64_t> grid_bdim) {
```

Comment:
I actually wonder if this is a more ... generic pattern. Hmmm....

Would it be possible to add a top level comment explaining the general rationale behind this pattern?

Reply:
Added [a comment](https://github.com/facebookresearch/functorch/pull/128/files#diff-e5658af94b6758a9726b8fade288818cb136f7dca498110a34973f2481b94c67R204-230) explaining the shapes in the 3 cases (for both 2d and 3d). Let me know if there's more needed to clarify it since it can be hard to read since both the input and grid have Hs and Bs

I was able to make it generic by basically copying what's done for binary op helper (the difference between the convolutions is that the non-cudnn version has more arguments about like padding and stuff). I'm not sure if we can make it more generic with conv (if that's what you were thinking or suggesting) since conv uses group convolutions to make the shapes work at least once and has to deal with bias while grid sampler never has a bias.